test/server/lib/GristJobs.ts

```diff
@@ -0,0 +1,149 @@
+        assert.equal(ct, 5);
+        assert.equal(defaultCt, 0);
+      });
+      await q.add('badd', {delta: 4});
+      await waitToPass(async () => {
+        assert.equal(ct, 5);
+        assert.equal(defaultCt, 1);
+      });
+    } finally {
+      await jobs.stop({obliterate: true});
+    }
+  });
+
+  it('can run delayed jobs', async function() {
+    const jobs: GristJobs = new GristBullMQJobs();
+    const q = jobs.queue();
+    try {
+      let ct = 0;
+      let defaultCt = 0;
+      q.handleName('add', async (job) => {
+        ct += job.data.delta;
+      });
+      q.handleDefault(async () => {
+        defaultCt++;
+      });
+      await q.add('add', {delta: 2}, {delay: 500});
+      assert.equal(ct, 0);
+      assert.equal(defaultCt, 0);
+      // We need to wait long enough to see the effect.
+      await delay(100);
+      assert.equal(ct, 0);
+      assert.equal(defaultCt, 0);
+      await delay(900);
+      assert.equal(ct, 2);
+      assert.equal(defaultCt, 0);
+    } finally {
+      await jobs.stop({obliterate: true});
+    }
+  });
+
+  it('can run repeated jobs', async function() {
+    const jobs: GristJobs = new GristBullMQJobs();
+    const q = jobs.queue();
+    try {
+      let ct = 0;
+      let defaultCt = 0;
+      q.handleName('add', async (job) => {
+        ct += job.data.delta;
+      });
+      q.handleDefault(async () => {
+        defaultCt++;
+      });
+      await q.add('add', {delta: 2}, {repeat: {every: 250}});
+      await q.add('badd', {delta: 2}, {repeat: {every: 100}});
+      assert.equal(ct, 0);
+      assert.equal(defaultCt, 0);
+      await delay(1000);
+      // allow for a lot of slop on CI
+      assert.isAtLeast(ct, 8 - 4);
+      assert.isAtMost(ct, 8 + 4);
+      assert.isAtLeast(defaultCt, 10 - 3);
+      assert.isAtMost(defaultCt, 10 + 3);
+    } finally {
+      await jobs.stop({obliterate: true});
+    }
+  });
+
+  it('can pick up jobs again', async function() {
+    // this test is only appropriate if we have an external queue.
+    if (!process.env.REDIS_URL) { this.skip(); }
+    const jobs1: GristJobs = new GristBullMQJobs();
+    const q = jobs1.queue();
+    try {
+      let ct = 0;
+      q.handleName('add', async (job) => {
+        ct += job.data.delta;
+      });
+      q.handleDefault(async () => {});
+      await q.add('add', {delta: 1}, {delay: 250});
+      await q.add('add', {delta: 1}, {delay: 1000});
+      await delay(500);
+      assert.equal(ct, 1);
+      await jobs1.stop();
+      const jobs2: GristJobs = new GristBullMQJobs();
+      const q2 = jobs2.queue();
+      try {
+        q2.handleName('add', async (job) => {
+          ct += job.data.delta * 2;
+        });
+        q2.handleDefault(async () => {});
+        await delay(1000);
+        assert.equal(ct, 3);
+      } finally {
+        await jobs2.stop({obliterate: true});
+      }
+    } finally {
+      await jobs1.stop({obliterate: true});
+    }
+  });
+});
+
```

Comment:
So long running tasks are a problem since BullMQ cancelation relies on non-open source features:
https://docs.bullmq.io/bullmq-pro/observables/cancelation

And for the Grist Labs SaaS, we are very restricted on how long we can wait for workers to shut down because of ECS/Fargate requirements.

This for me is related to the big operational question mark about how to properly drain queues as workers are replaced, which we will have to deal with. There will be more work to do here.